docs/design/upgrade.md

```diff
@@ -0,0 +1,203 @@
+
+```typescript
+// Rename a class to another name. Need to pass a symbol as input, as just
+// a string is not enough.
+declare function renameClass(project: Angular2Project,
+                             from: ts.Symbol, to: string): Observable<Change>;
+// Upgrade a package in NPM.
+declare function upgradePackage(project: Angular2Project,
+                                name: string, to: string): Observable<Change>;
+// Change a configuration in angular-cli config.
+declare function configSet(project: Angular2Project, key: string,
+                           value: any): Observable<Change>;
+// Remove a file.
+declare function deleteFile(project: Angular2Project,
+                            file: string): Observable<Change>;
+// Scaffold a new blueprints.
+declare function scaffold(project: Angular2Project, name: string,
+                          arguments: {[name: string]: string}): Observable<Change>;
+```
+
+Some higher level changes examples:
+
+```typescript
+// Add a new Import if the import isn't already in the file.
+class AddImportChange implements Change {
+  constructor(file: string, symbolName: string, importPath: string);
+}
+```
+
+# Upgrade Command
+A new command must be added to the CLI;
+
+    ng upgrade [options]
+    Upgrade the project to a newer version.
+
+    Options:
+      --package=<name>  The package to upgrade. By default, upgrade every packages
+                        that are outdated.
+      --to=<semver>     Move to a specific version of Angular. Only valid with
+                        --package and --dry-run.
+                        Default: latest.
+      --from=<semver>   Specify the version the project is currently using. Only
+                        valid with --package and --dry-run.
+                        Default: the installed version is used.
+      --commit=<bool>   Whether the tool should automatically create a commit.
+                        Default: false.
+      --message=<msg>   The commit message to use. Only valid with --commit.
+      --dry-run         Only output the commands and refactor steps that would be run,
+                        but do nothing.
+                        Default: false.
+
+## Process
+The command will execute the following tasks in order:
+
+1. If the `angular-cli` is not the latest version, return with an error message suggesting an upgrade. This step is necessary even if the version is supported as a upgrade, because updates to the upgrade steps might have occurred.
+1. If the project is in a git repository, verify that there are no differences between the current ref and the latest origin/master. If differences exist (or status is not clean), exit early.
+1. If `--package` is passed:
+    1. Check that that package is installed.
+    1. If `--from` is passed, check that `--dry-run` is passed too.
+    1. If `--to` is passed, check that `--dry-run` is passed too.
+1. Otherwise, list all packages that need updating.
```

Comment:
Not sure it ever makes sense to update more than one package at a time. There's peerDependencies to consider - at an 1-to-1 level, and at a n-to-n level.

At an 1-to-1 level you can, for instance, have `angular2@2.0.0` and `rxjs@5.0.0`.    We might want to upgrade both, but `angular2@2.0.1` might require `rxjs@5.0.0` or even `rxjs@5.0.2`. Thus we wouldn't be able to do both at the same time.

Even more complicated is n-to-n, where packages might have incompatible peerDeps.

I understand the need to be able to update several packages at a time (again because of peer dependencies), but it should be targeted: I want to upgrade `angular2` and *because* of that I also have to upgrade `rxjs` and `zones`.